### PR TITLE
[tt-crank] fused SDPA training via ttml sdpa_fw/sdpa_bw

### PR DESCRIPTION
**Pull request by agobeljicTT**
Training through `F.scaled_dot_product_attention` no longer decomposes into matmul + softmax. With grad live, the overrideable op lowers to the `sdpa_fw` / `sdpa_bw` `ttcore.composite`s, promoted to the ttml kernels at `OPT_LEVEL >= 1` and inlined as plain TTIR at `OPT_LEVEL 0`. Inference keeps `ttir.scaled_dot_product_attention`. One implementation (`build_sdpa_fw` / `build_sdpa_bw`) serves eager, compile and DTensor (`Shard(0)` / `Shard(1)`).

Fused when: bf16, 4-D, `Sq == Sk`, `S % 32 == 0`, causal or maskless or one `S x S` bool mask shared by all batches and heads (leading dims of 1), MHA or GQA, any `scale` / head_dim. Otherwise the math decomposition with a one-time warning; `dropout_p > 0` raises.

Perf on top of #9302, Llama 3.2 1B SFT (OPT_LEVEL 1, bs=1, sl=1024, 1 chip): step time 0.789 s to 0.641 s, MFU 6.23% to 7.66%.

Tests: `tests/python/test_sdpa.py`, `tests/python/models/test_attention_training.py`.

### CODE REVIEW

File: tt-crank/python/tt_crank/torch/_compile.py [hunk 1]
```diff
@@ -126,6 +127,20 @@ def decorator(fn):
...
+    fn._needs_node = True
+    return fn
+
+
+def _used_output_slots(node) -> set[int]:
+    return {
+        u.args[1]
+        for u in node.users
+        if u.op == "call_function" and u.target is operator.getitem
+    }
```
[the remark this answers is not in the document]
Comment (agobeljicTT) [reply tt-crank/python/tt_crank/torch/_compile.py:141]: not an issue. dynamo unpacks a tuple slice into per-element getitem nodes with int indices before the graph reaches the backend, a slice never shows up as a getitem arg on an aten tuple output. the lse test uses exactly `[:2]` and asserts `ttnn.sdpa_fw` is in the ttnn ir, it passes. slices are also hashable since python 3.12.

File: tt-crank/src/torch/ops/sdpa.cpp [hunk 2]
```diff
@@ -44,17 +46,39 @@ namespace tt::crank::torch_backend {
...
-    // has no meta kernel. MATH decomposes to differentiable primitives instead, so
-    // training keeps working at the cost of the single fused op.
-    if (at::GradMode::is_enabled() && (query.requires_grad() || key.requires_grad() || value.requires_grad())) {
-        TORCH_WARN_ONCE("tt-crank sdpa: inputs require grad, so SDPA is using the math decomposition "
-                        "instead of the fused tt kernel. For inference, run under torch.no_grad() or "
-                        "torch.inference_mode() to get the fused op.");
+                            const std::optional<at::Tensor> &attn_mask, double dropout_p, bool, std::optional<double>,
+                            bool) {
+    const bool training =
+        at::GradMode::is_enabled() && (query.requires_grad() || key.requires_grad() || value.requires_grad());
```
[the remark this answers is not in the document]
Comment (agobeljicTT) [reply tt-crank/src/torch/ops/sdpa.cpp:75]: premise is off. the choice stub is called by the sdpa composite with the dtensor wrapper, where requires_grad is intact, so training is detected and unsupported shapes go to math. only the overrideable kernel gets redispatched on local tensors, which is why it gates on grad mode alone. checked on device: dtensor shard(0), eager, s=48 picks math (and hits the pre-existing dtensor causal mask decomposition error, not the fused backward). the empty logsumexp branch is only reached for inference under grad mode with no requires_grad, where no backward is recorded.

**Comment by agobeljicTT** [comment]
splitting PR into multiples
